onnxruntime/core/providers/cpu/tensor/scatter_nd.cc

```diff
@@ -168,54 +186,131 @@ Status ScatterNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) co
+struct Func_Add_ND<MLFloat16> {
+  void operator()(MLFloat16*, const MLFloat16*, uint64_t) const {
+    ORT_NOT_IMPLEMENTED("CPU execution provider: MLFloat16 data type is not supported with ScatterND opset 16 when reduction is 'add'.");
+    }
+};
+
+template<>
+struct Func_Add_ND<BFloat16> {
+  void operator()(BFloat16*, const BFloat16*, uint64_t) const {
+    ORT_NOT_IMPLEMENTED("CPU execution provider: BFloat16 data type is not supported with ScatterND opset 16 when reduction is 'add'.");
+    }
+};
 
-Status ScatterND::ScatterString(const Prepare& p, concurrency::ThreadPool* tp) const {
-  if (!utils::HasType<EnabledScatterNDDataTypes, std::string>()) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "String data type is not supported in this build.");
+template <class T>
+struct Func_Mul_ND {
+  void operator()(T* a, const T* b, uint64_t element_to_copy) const {
+    while (element_to_copy-- > 0)
+      (*a++) *= (*b++);
   }
+};
 
-  auto lambda = [&](int64_t i) {
-    for (int64_t j = 0; j < static_cast<int64_t>(p.element_to_copy); ++j) {
-      p.output_str_base[p.element_offsets[i] + j] = p.input_str_base[i * p.element_to_copy + j];
+template<>
+struct Func_Mul_ND<bool> {
+  void operator()(bool* a, const bool* b, uint64_t element_to_copy) const {
+    while (element_to_copy-- > 0)
+      (*a++) &= (*b++);
+  }
+};
+
+template<>
+struct Func_Mul_ND<std::string> {
+  void operator()(std::string*, const std::string*, uint64_t) const {
+    ORT_NOT_IMPLEMENTED("CPU execution provider: string data type is not supported with ScatterND opset 16 when reduction is 'mul'.");
+  }
+};
+
+template<>
+struct Func_Mul_ND<MLFloat16> {
+  void operator()(MLFloat16*, const MLFloat16*, uint64_t) const {
+    ORT_NOT_IMPLEMENTED("CPU execution provider: MLFloat16 data type is not supported with ScatterND opset 16 when reduction is 'mul'.");
     }
-  };
-  concurrency::ThreadPool::TryParallelFor(tp, p.element_offsets.size(), static_cast<double>(p.element_to_copy),
-                                          [&lambda](ptrdiff_t first, ptrdiff_t last) {
-                                            for (int i = static_cast<int>(first), end = static_cast<int>(last); i < end; ++i) {
-                                              lambda(i);
-                                            }
-                                          });
-  return Status::OK();
-}
+};
 
+template<>
+struct Func_Mul_ND<BFloat16> {
+  void operator()(BFloat16*, const BFloat16*, uint64_t) const {
+    ORT_NOT_IMPLEMENTED("CPU execution provider: BFloat16 data type is not supported with ScatterND opset 16 when reduction is 'mul'.");
+    }
+};
+
+template <typename TData>
+struct ScatterNDDispatchTarget {
+  Status operator()(OpKernelContext* context, concurrency::ThreadPool* tp, const std::string &reduction) const {
+    Prepare<TData> prepare;
+    ORT_RETURN_IF_ERROR(PrepareForCompute(context, prepare));
+
+    auto lambda = [&](int64_t i) {
+      if (reduction == "add") {
```

Comment:
minor nit (for future enhancement): no need to repeat this test for every `i` in the loop. We can pull the conditional out of the lambda as below:
```cpp
   auto lambda = (reduction == "add") ?
                                (lambda for add)  : 
                           (reduction == "mul") ?
                                (lambda for mul) : 
                                (lambda for none);
```